ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/trielog/TrieLogPruner.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright contributors to Hyperledger Besu.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.hyperledger.besu.ethereum.bonsai.trielog;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage;
+import org.hyperledger.besu.ethereum.chain.Blockchain;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.ProcessableBlockHeader;
+
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Stream;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.TreeMultimap;
+import org.apache.tuweni.bytes.Bytes32;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TrieLogPruner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TrieLogPruner.class);
+
+  private final int pruningLimit;
+  private final int loadingLimit;
+  private final BonsaiWorldStateKeyValueStorage rootWorldStateStorage;
+  private final Blockchain blockchain;
+  private final long numBlocksToRetain;
+  private final boolean requireFinalizedBlock;
+
+  private final Multimap<Long, Hash> trieLogBlocksAndForksByDescendingBlockNumber =
+      TreeMultimap.create(Comparator.reverseOrder(), Comparator.naturalOrder());
+
+  public TrieLogPruner(
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final Blockchain blockchain,
+      final long numBlocksToRetain,
+      final int pruningLimit,
+      final boolean requireFinalizedBlock) {
+    this.rootWorldStateStorage = rootWorldStateStorage;
+    this.blockchain = blockchain;
+    this.numBlocksToRetain = numBlocksToRetain;
+    this.pruningLimit = pruningLimit;
+    this.loadingLimit = pruningLimit; // same as pruningLimit for now
+    this.requireFinalizedBlock = requireFinalizedBlock;
+  }
+
+  public void initialize() {
+    preloadQueue();
+  }
+
+  private void preloadQueue() {
+    LOG.atInfo()
+        .setMessage("Loading first {} trie logs from database...")
+        .addArgument(loadingLimit)
+        .log();
+    try {
+      final Stream<byte[]> trieLogKeys = rootWorldStateStorage.streamTrieLogKeys(loadingLimit);
```

Comment:
Think this should be done in a try-with-resources so that we close the stream

Reply:
done https://github.com/hyperledger/besu/pull/6026/commits/13f0040867102e9c6b0a627b3d61eb792769d0f0